components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +295,245 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
 
         return oAuthClientAuthenticator.canAuthenticate(request, bodyContentMap, oAuthClientAuthnContext);
     }
+
+    /**
+     * Validate whether the client authentication method of the request is registered for the application.
+     *
+     * @param request                Http servlet request.
+     * @param contentParams          Map of request body params.
+     * @return Whether the client authentication method of the request is registered for the application.
+     * @throws OAuthClientAuthnException
+     */
+    private boolean isRegisteredClientAuthMethod(HttpServletRequest request, Map<String, List> contentParams)
+            throws OAuthClientAuthnException {
+
+        String registeredClientAuthMethod = retrieveRegisteredAuthMethod(extractClientId(request, contentParams));
+
+        if (registeredClientAuthMethod.equals(OAuthConstants.NOT_APPLICABLE)) {
+            return false;
+        }
+
+        // There can be multiple registered client auth methods
+        if (!(registeredClientAuthMethod.contains(retrieveRequestAuthMethod(request, contentParams)))) {
+            throw new OAuthClientAuthnException("Request does not follow the registered token endpoint auth " +
+                    "method", OAuth2ErrorCodes.INVALID_REQUEST);
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Obtain the client authentication method of the request.
+     *
+     * @param request                Http servlet request.
+     * @param contentParams          Map of request body params.
+     * @return Client authentication method of the request.
+     * @throws OAuthClientAuthnException
+     */
+    public String retrieveRequestAuthMethod(HttpServletRequest request, Map<String, List> contentParams) throws
+            OAuthClientAuthnException {
+
+        try {
+            if (isPrivateKeyJWTAuthentication(request)) {
+                log.debug("Validating request with JWT client authentication method");
+                return OAuthConstants.PRIVATE_KEY_JWT;
+            } else if (isMTLSAuthentication(request, contentParams)) {
+                log.debug("Validating request with MTLS client authentication method");
+                return OAuthConstants.TLS_CLIENT_AUTH;
+            }
+            return "INVALID_AUTH";
```

Comment:
Move to a constant

Reply:
Fixed in 36cde69f01a0865dc525ff8c963f66e0af53cf89